src/main/java/net/minecraftforge/client/model/ModelFluid.java

```diff
@@ -115,53 +120,48 @@ public IModel loadModel(ResourceLocation modelLocation)
                 {
-                    cornerRound[i] = (int)(key & 0x3FF);
+                    cornerRound[i] = (int) (key & 0x3FF);
                     key >>>= 10;
                 }
-                int flowRound = (int)(key & 0x7FF) - 1024;
-                return new BakedFluid(transformation, transforms, format, color, still, flowing, gas, statePresent, cornerRound, flowRound);
+                int flowRound = (int) (key & 0x7FF) - 1024;
+                key >>>= 11;
+                boolean[] overlaySides = new boolean[4];
+                for (int i = 0; i < 4; i++)
+                {
+                    overlaySides[i] = (key & 1) != 0;
+                    key >>>= 1;
+                }
+                return new BakedFluid(transformation, transforms, format, color, still, flowing, overlay, gas, statePresent, cornerRound, flowRound, overlaySides);
             }
         });
 
-        private final Optional<TRSRTransformation> transformation;
-        private final ImmutableMap<TransformType, TRSRTransformation> transforms;
-        private final VertexFormat format;
-        private final int color;
-        private final TextureAtlasSprite still, flowing;
-        private final boolean gas;
-        private final EnumMap<EnumFacing, List<BakedQuad>> faceQuads;
-
-        public BakedFluid(Optional<TRSRTransformation> transformation, ImmutableMap<TransformType, TRSRTransformation> transforms, VertexFormat format, int color, TextureAtlasSprite still, TextureAtlasSprite flowing, boolean gas, Optional<IExtendedBlockState> stateOption)
+        public CachingBakedFluid(Optional<TRSRTransformation> transformation, ImmutableMap<TransformType, TRSRTransformation> transforms, VertexFormat format, int color, TextureAtlasSprite still, TextureAtlasSprite flowing, Optional<TextureAtlasSprite> overlay, boolean gas, Optional<IExtendedBlockState> stateOption)
         {
-            this(transformation, transforms, format, color, still, flowing, gas, stateOption.isPresent(), getCorners(stateOption), getFlow(stateOption));
+            super(transformation, transforms, format, color, still, flowing, overlay, gas, stateOption.isPresent(), getCorners(stateOption), getFlow(stateOption), getOverlay(stateOption));
         }
 
         private static int[] getCorners(Optional<IExtendedBlockState> stateOption)
         {
-            int[] cornerRound = new int[]{0, 0, 0, 0};
-            if(stateOption.isPresent())
+            int[] cornerRound = {0, 0, 0, 0};
+            if (stateOption.isPresent())
             {
                 IExtendedBlockState state = stateOption.get();
-                for(int i = 0; i < 4; i++)
+                for (int i = 0; i < 4; i++)
                 {
                     Float level = state.getValue(BlockFluidBase.LEVEL_CORNERS[i]);
-                    cornerRound[i] = Math.round((level == null ? 7f / 8 : level) * 768);
+                    cornerRound[i] = Math.round((level == null ? 8f / 9f : level) * 864);
```

Comment:
Document your magic numbers.

Reply:
For reference here: 864 is 9/8ths of 768, so that 768 / 8 = 864 / 9 = 96.

The choice of value here is a compromise between making better use of the 2^10 value range available within the cache key and being able to represent most common values exactly.